### ISSUE TITLE
Not able to reproduce reported results for adaptive O-CNN classification and autoencoder

### ISSUE THREAD

**Issue by jeichelbaum** [microsoft/O-CNN#111]
### EDIT: managed to reproduce classification accuracy and autoencoder chamfer distance as reported in paper

I am able to reliable reproduce the results for the full O-CNN classification experiment, but not for the adaptive O-CNN classification and autoencoder experiment. I gathered the results you [reported](https://arxiv.org/pdf/1809.07917.pdf) and the results I was able to reproduce in a table below:

|   | depth 5 | depth 6 | depth 7 | 
| ------------- | ------------- |  ------------- |  ------------- |
| **Reported AO-CNN classification accuracy**  | 90.5%  | 90.4%  | 90.0%  |
| Caffe AO-CNN accuracy  | 85.12% | 85.68% | 86.18% | 
| Tensorflow AO-CNN accuracy  | 83.38% | 83.72% | 84.31% | 
| **Reported autoencoder avg. Chamfer dist.**  | - | - | 1.44 | 
| Reported autoencoder avg. Chamfer dist.  | - | - | 1.77 | 

### Classification experiment
I am using the ModelNet40 point dataset you provide and followed the instructions exactly as you posted them in docs/classification. I used both the Caffe and Tensorflow implementation to test this, but didn't get accuracy close to your results. The model I used is the same you uploaded in [caffe/experiments/aocnn_m40_5.prototxt](https://github.com/microsoft/O-CNN/blob/master/caffe/experiments/aocnn_m40_5.prototxt) and [tensorflow/script/configs/cls_octree.yaml](https://github.com/microsoft/O-CNN/blob/master/tensorflow/script/configs/cls_octree.yaml) I attached a visualization of the Caffe model for convenience: 
![train](https://user-images.githubusercontent.com/4103504/100170781-b3ceaa00-2ec6-11eb-9a6e-8e51e94d5a6c.png)

Going through the Tensorflow code spawned some question: 

1. Does the Tensorflow classifier implementation support adaptive octrees as input? 
2. Did you use different parameters to construct the octrees other than those mentioned in [caffe/experiments/dataset.py](https://github.com/microsoft/O-CNN/blob/master/caffe/experiments/dataset.py)?
3. What signal size did you use in the adaptive O-CNN classifier experiment? The config files use signal_size=3 (just normal), but from your paper I get the idea that you might have used signal_size = 4 (normal + displacement). I already tried the experiment with both signal_sizes but without success.

### Autoencoder experiment
Again I used the ShapeNet point cloud you provided in combination with the provided [caffe/experiments/ae_7_4.train.prototxt](https://github.com/microsoft/O-CNN/blob/master/caffe/experiments/ae_7_4.train.prototxt) and executed each step as you describe it in the docs section, but my results are still way off. 

#### My big question is:
Do you have any pointers for me, how I can get closer to reproducing your experiments? Or would you be so kind to share the trained models for both the Caffe adaptive OCNN classifier and autoencoder?

Thanks in advance

**Comment by wang-ps** [comment]
1. For the adaptive o-cnn, the basic component is already implemented in our Tensorflow-based implementation. However, I have not tested the adaptive o-cnn on tensorflow. 

2. I have updated the `dataset.py` and fixed one parameter. Please pull the latest code and have a try. With depth-5 adaptive o-cnn, the classification accuracy is 89.4% before voting. After using the voting strategies, such as orientation pooling as mentioned in our paper, the performance will increase to about 90.5%.   I have just run the experiment on my own PC today. The trained log and the last caffemodel can be downloaded [here](https://github.com/microsoft/O-CNN/files/5600725/aocnn_m40_5.zip). Other results will be released soon.

3. In the classification experiments, we use only the first 3 channals. According to my own experiments, with 4-channel signals, the testing accuracy may drop about 0.2, probably due to overfitting. In the autoencoder experiments, 4-channel signals are used.

**Comment by jeichelbaum** [comment]
Thanks for your reply and clarification! Your commit fixed the adaptive O-CNN classification experiment for me and now I get around 90% classification accuracy. Apparently I was using two wrong parameters for the build_octree tool: axis=y and displacment=1 instead of the correct parameters axis=z and displacement=0

I will test the adaptive autoencoder next and will post an update as soon as I have results.

**Comment by jeichelbaum** [comment]
I implemented voting to see if I am able to reproduce the accuracy for the adaptive O-CNN classification experiment. I am seeing some improvements, but I am still not able to completely reproduce your results. 

|   | depth 5 | depth 6 | depth 7 | 
| ------------- | ------------- |  ------------- |  ------------- |
| **Reported AO-CNN classification accuracy**  | 90.5%  | 90.4%  | 90.0%  |
| Caffe AO-CNN accuracy voting  | 89.7% | 89.2% | 89.1% |

**Comment by wang-ps** [comment]
1. Please compare your training log with the one I uploaded in my last post.  And you can run several times and test the accuracies. 

2. With a simple average voting strategy, the accuracy may increase by about 0.5%. With orientation pooling (for the detailed implementation please refer to Section 4.4 of [this paper](https://arxiv.org/abs/1604.03265), the accuracy may increase by about 1.0%.

**Comment by jeichelbaum** [comment]
1. I ran the experiment multiple times, but it doesn't change much. The behavior over time is consistent with your training log, but I always seem to miss around 0.3% in accuracy. In my opinion this is tolerable and a lot better than the 85.1% accuracy of my initial training round.

2. Thank you for clarifying and walking me through the process to reproduce your results! Orientation pooling explains the gap in accuracy and concludes the classification experiment for me. 

3. I am still working on the Autoencoder experiment and will post an update soon

**Comment by wang-ps** [comment]
I have pushed the latest code to the master branch. 
I automated the process to do the autoencoder experiment with python. I did the experiment last week, the results are reproducible. The trained log, model, and resulting chamfer distance can be downloaded [here](https://www.dropbox.com/s/nta8tnior85j6zn/aocnn_ae.zip?dl=0).
